reactive/calico.py

```diff
@@ -323,22 +340,39 @@ def ready():
 
 @when('config.changed.registry-credentials')
 def registry_credentials_changed():
-    status_set('maintenance', 'Applying registry credentials')
-    encoded_creds = hookenv.config('registry-credentials')
+    remove_state('calico.image.pulled')
 
-    if not encoded_creds:
-        log('no registry-credentials, skipping docker config')
-        return
 
+@when('calico.ctl.ready')
+@when_not('calico.image.pulled')
+def pull_calicoctl_image():
+    status_set('maintenance', 'Pulling calicoctl image')
+    registry = hookenv.config('registry') or DEFAULT_REGISTRY
+    images = [CALICOCTL_IMAGE, CNXNODE_IMAGE]
+
+    encoded_creds = hookenv.config('registry-credentials')
     creds = b64decode(encoded_creds).decode('utf-8')
-    creds = json.loads(creds)
-    for server_name, server in creds['auths'].items():
-        auth = server['auth']
+    if creds:
+        creds = json.loads(creds)
+
+    if not creds or not creds.get('auths') or \
+            registry not in creds.get('auths'):
+        for image in images:
+            CTL.pull(
+                registry + image,
+            )
```

Comment:
If I understand correctly, over in https://github.com/charmed-kubernetes/layer-calico/pull/36, you've added the calico-node-image resource as a workaround for the fact that `CTL.pull` does not respect the proxy configuration of the containerd subordinate. Correct?

So we'll need a similar workaround for the images here. But, using charm resources as a workaround here is more problematic because the cnx-node and calicoctl images used by this charm are protected proprietary images. We can't legally host them in the charm store. Which means the user would have to be responsible for manually attaching the images.

Reply:
From IRC, the plan is to ship this with a zero-byte resource in the charm store, and provide documentation for users to attach their own image as needed for offline deployments. Works for me.